dbms/src/Common/Allocator.cpp

```diff
@@ -114,15 +115,18 @@ void Allocator<clear_memory_>::free(void * buf, size_t size)
 template <bool clear_memory_>
 void * Allocator<clear_memory_>::realloc(void * buf, size_t old_size, size_t new_size, size_t alignment)
 {
-#if !defined(__APPLE__) && !defined(__FreeBSD__)
     if (old_size < MMAP_THRESHOLD && new_size < MMAP_THRESHOLD && alignment <= MALLOC_MIN_ALIGNMENT)
     {
         CurrentMemoryTracker::realloc(old_size, new_size);
 
-        buf = ::realloc(buf, new_size);
+        void * new_buf = ::realloc(buf, new_size);
 
-        if (nullptr == buf)
+        if (nullptr == new_buf)
+        {
+            ::free(buf);
```

Comment:
Double free?

Reply:
realloc:
On failure, returns a null pointer. The original pointer ptr remains valid and may need to be deallocated with free() or realloc().

so we need to free if realloc returns nullptr

Reply:
It means that realloc is "transactional". If it fails, it leaves the memory in previous state.
And it must be freed by user, exactly as in case when there was no realloc.